### PR TITLE
[GSoC-2017] built-in reports implemented with SQL data definitions

### PR DESCRIPTION
**Pull request by JudeNiroshan**
Reports have been implemented using SQL data definitions. 

- [x] OpenMRS code fomatter used
- [x] mvn clean install successful
- [x] all implemented reports are available once the module is successfully deployed to the server

https://issues.openmrs.org/browse/RA-1257

### CODE REVIEW

**Comment by dkayiwa** [comment]
@JudeNiroshan it would be easier to review these if they are squashed into one commit.

File: api/src/main/java/org/openmrs/module/referenceapplication/builtinreports/reports/ListOfDiagnosis.java [hunk 1]
```diff
@@ -0,0 +1,92 @@
+package org.openmrs.module.referenceapplication.builtinreports.reports;
```
Comment (rkorytkowski) [inline api/src/main/java/org/openmrs/module/referenceapplication/builtinreports/reports/ListOfDiagnosis.java]: Please include OpenMRS license header in all new files, see https://github.com/openmrs/openmrs-core/blob/master/api/src/main/java/org/openmrs/Concept.java#L1-L9

Move the classes to org.openmrs.module.referenceapplication.reports (skip the builtinreports)

File: api/src/main/java/org/openmrs/module/referenceapplication/builtinreports/reports/ListOfDiagnosis.java [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+import org.openmrs.module.reporting.report.definition.ReportDefinition;
+import org.openmrs.module.reporting.report.manager.BaseReportManager;
+import org.springframework.stereotype.Component;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Created by Jude on 6/3/2017.
+ */
```
Comment (rkorytkowski) [inline api/src/main/java/org/openmrs/module/referenceapplication/builtinreports/reports/ListOfDiagnosis.java]: We do not include Created by... in comments by convention. Please remove it from all newly created classes. No worries, your contribution is visible in the git history.

File: api/src/main/java/org/openmrs/module/referenceapplication/ReferenceApplicationActivator.java [hunk 3]
```diff
@@ -105,6 +107,8 @@ public void started() {
             setupTagLocation(ReferenceMetadataConstants.VISIT_LOCATION_TAG_UUID);
 	        setupHtmlForms();
 	        setupHL7ProcessingTask(schedulerService);
+
+			ReportManagerUtil.setupAllReports(ReportManager.class);
```
Comment (rkorytkowski) [inline api/src/main/java/org/openmrs/module/referenceapplication/ReferenceApplicationActivator.java]: Could you please add a comment saying that it sets up reports from org.openmrs.module.referenceapplication.reports?

File: api/src/main/resources/moduleApplicationContext.xml [hunk 4]
```diff
@@ -18,6 +18,6 @@
   		    http://www.springframework.org/schema/util/spring-util-3.0.xsd">
 
 	<!-- Add here beans related to the API context -->
-
+	<context:component-scan base-package="org.openmrs.module.referenceapplication.builtinreports" />
```
Comment (rkorytkowski) [inline api/src/main/resources/moduleApplicationContext.xml]: Is this necessary? Could you please try removing it? If it is needed, then it should look for components in the whole module i.e. org.openmrs.module.referenceapplication.

File: pom.xml [hunk 5]
```diff
@@ -231,6 +233,14 @@
 	            <version>${htmlformentryVersion}</version>
 	            <scope>provided</scope>
 	        </dependency>
+
+	        <dependency>
+		        <groupId>org.openmrs.module</groupId>
+		        <artifactId>reporting-api-2.0</artifactId>
+		        <version>${reportingVersion}</version>
+		        <scope>provided</scope>
+	        </dependency>
```
Comment (rkorytkowski) [inline pom.xml]: You also need to add reporting to required modules in https://github.com/openmrs/openmrs-module-referenceapplication/blob/master/omod/src/main/resources/config.xml
It is to make sure the reporting module is installed and started before the referenceapplication module.

**Pull request closed without merging**